File: services/control/internal/handler.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// /v1/control/invoke handler: parses JSON, authenticates the bearer, blocks JTI replay, rate-limits, dispatches, and audits every outcome.

package internal

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"net/http"
	"time"

	"github.com/rs/zerolog"
)

const maxBodyBytes = 64 * 1024

func InvokeHandler(auth *Authenticator, disp *Dispatcher, sink EventSink, rate *RateLimiter, replay *ReplayCache, log zerolog.Logger) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		reqID := newRequestID()
		w.Header().Set("X-Request-Id", reqID)

		if r.Method != http.MethodPost {
			writeJSON(w, http.StatusMethodNotAllowed, Response{Error: "method not allowed"})
			return
		}
		ctx, cancel := context.WithTimeout(r.Context(), 30*time.Second)
		defer cancel()

		claims, err := auth.Verify(ctx, r.Header.Get("Authorization"))
		if err != nil {
			sink.Emit(AuditEvent{At: time.Now().UTC(), Subject: "anonymous", Decision: "deny", Reason: "auth: " + err.Error(), RequestID: reqID})
			writeJSON(w, http.StatusUnauthorized, Response{Error: "unauthorized"})
			return
		}
		exp := time.Time{}
		if claims.ExpiresAt != nil {
			exp = claims.ExpiresAt.Time
		}
		if !replay.Mark(claims.ID, exp) {
			sink.Emit(AuditEvent{At: time.Now().UTC(), Subject: claims.Subject, JTI: claims.ID, Decision: "deny", Reason: "replay", RequestID: reqID})
			writeJSON(w, http.StatusUnauthorized, Response{Error: "token replay"})
			return
		}
		if !rate.Allow(claims.Subject) {
			sink.Emit(AuditEvent{At: time.Now().UTC(), Subject: claims.Subject, JTI: claims.ID, Decision: "deny", Reason: "rate limited", RequestID: reqID})
			writeJSON(w, http.StatusTooManyRequests, Response{Error: "rate limited"})
			return
		}
		r.Body = http.MaxBytesReader(w, r.Body, maxBodyBytes)
		var req Request
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			sink.Emit(AuditEvent{At: time.Now().UTC(), Subject: claims.Subject, JTI: claims.ID, Decision: "deny", Reason: "invalid json", RequestID: reqID})
			writeJSON(w, http.StatusBadRequest, Response{Error: "invalid json"})
			return
		}
		result, err := disp.Dispatch(ctx, req)
		event := AuditEvent{
			At:        time.Now().UTC(),
			Subject:   claims.Subject,
			JTI:       claims.ID,
			Command:   req.Command,
			Sub:       req.Subcommand,
			Decision:  "allow",
			RequestID: reqID,
		}
		if err != nil {
			event.Decision = "deny"
			event.Reason = err.Error()
			sink.Emit(event)
			switch {
			case errors.Is(err, ErrDenied):
				writeJSON(w, http.StatusBadRequest, Response{Error: err.Error()})
			case errors.Is(err, ErrUnsupported):
				writeJSON(w, http.StatusNotImplemented, Response{Error: err.Error()})
			default:
				log.Err(err).Str("cmd", req.Command).Msg("upstream error")
				writeJSON(w, http.StatusBadGateway, Response{Error: "upstream error"})
			}
			return
		}
		sink.Emit(event)
		writeJSON(w, http.StatusOK, Response{OK: true, Result: result})
	})
}

func writeJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(body)
}

func newRequestID() string {
	var b [16]byte
	_, _ = rand.Read(b[:])
	return hex.EncodeToString(b[:])
}

File: apps/control/src/handler.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// /v1/control/invoke handler: rate-limits, authenticates, blocks JTI replay, and dispatches through the shared engine.

import type { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify'
import { dispatch, DispatchError, type DispatchContext, type FlagMap, type Principal } from '@caracalai/engine'
import { Authenticator, AuthError } from './auth.js'
import { newRequestId, type EventSink } from './audit.js'
import type { Replay } from './replay.js'
import type { RateLimiter } from './ratelimit.js'
import type { ControlGate } from './gate.js'

const MAX_BODY_BYTES = 64 * 1024

interface InvokeBody {
  command?: unknown
  subcommand?: unknown
  flags?: unknown
}

interface RouteRateLimit {
  readonly max: number
  readonly timeWindow: number
}

export interface InvokeDeps {
  auth: Authenticator
  replay: Replay
  rate: RateLimiter
  routeRateLimit: RouteRateLimit
  sink: EventSink
  ctx: DispatchContext
  gate: ControlGate
}

export function registerInvokeRoute(app: FastifyInstance, deps: InvokeDeps): void {
  app.post('/v1/control/invoke', {
    bodyLimit: MAX_BODY_BYTES,
    config: { rawBody: false, rateLimit: deps.routeRateLimit },
  }, (req, reply) => handle(req, reply, deps))
}

async function handle(req: FastifyRequest, reply: FastifyReply, deps: InvokeDeps): Promise<void> {
  const requestId = newRequestId()
  reply.header('x-request-id', requestId)
  if (!deps.gate.enabled()) {
    await deps.sink.emit({
      at: new Date(), subject: 'anonymous', jti: '',
      decision: 'deny', reason: 'control disabled', requestId,
    })
    return reply.code(503).send({ error: 'control disabled' })
  }

  let claims
  try {
    claims = await deps.auth.verify(req.headers.authorization)
  } catch (err) {
    await deps.sink.emit({
      at: new Date(), subject: 'anonymous', jti: '',
      decision: 'deny', reason: 'auth: ' + describe(err), requestId,
    })
    return reply.code(401).send({ error: 'unauthorized' })
  }

  if (!(await deps.replay.mark(claims.jti, claims.exp))) {
    await deps.sink.emit({
      at: new Date(), zoneId: claims.zoneId, clientId: claims.clientId,
      subject: claims.sub, jti: claims.jti, decision: 'deny', reason: 'replay', requestId,
    })
    return reply.code(401).send({ error: 'token replay' })
  }
  if (!deps.rate.allow(claims.sub)) {
    await deps.sink.emit({
      at: new Date(), zoneId: claims.zoneId, clientId: claims.clientId,
      subject: claims.sub, jti: claims.jti, decision: 'deny', reason: 'rate limited', requestId,
    })
    return reply.code(429).send({ error: 'rate limited' })
  }

  const body = req.body as InvokeBody | null
  const command = typeof body?.command === 'string' ? body.command : ''
  const subcommand = typeof body?.subcommand === 'string' ? body.subcommand : ''
  const flags = (body?.flags && typeof body.flags === 'object' && !Array.isArray(body.flags))
    ? body.flags as FlagMap
    : undefined

  const principal: Principal = {
    kind: 'remote',
    subject: claims.sub,
    zoneId: claims.zoneId,
    clientId: claims.clientId,
    scopes: claims.scope.split(/\s+/).filter((s) => s.length > 0),
  }

  try {
    const result = await dispatch({ command, subcommand, flags }, principal, deps.ctx)
    await deps.sink.emit({
      at: new Date(), zoneId: claims.zoneId, clientId: claims.clientId,
      subject: claims.sub, jti: claims.jti, command, subcommand,
      decision: 'allow', requestId,
    })
    return reply.code(200).send({ ok: true, result })
  } catch (err) {
    const reason = describe(err)
    await deps.sink.emit({
      at: new Date(), zoneId: claims.zoneId, clientId: claims.clientId,
      subject: claims.sub, jti: claims.jti, command, subcommand,
      decision: 'deny', reason, requestId,
    })
    if (err instanceof DispatchError) {
      if (err.code === 'denied') return reply.code(403).send({ error: 'denied' })
      if (err.code === 'invalid') return reply.code(400).send({ error: 'invalid request' })
      return reply.code(501).send({ error: 'unsupported' })
    }
    req.log.error({ command }, 'upstream error: ' + reason)
    return reply.code(502).send({ error: 'upstream error' })
  }
}

function describe(err: unknown): string {
  if (err instanceof AuthError) return err.message
  if (err instanceof Error) return err.message
  return String(err)
}
